src/systems/buoyancy/Buoyancy.cc

```diff
@@ -62,18 +73,231 @@ class ignition::gazebo::systems::BuoyancyPrivate
   /// pose frame is left undefined because this function currently returns
   /// a constant value, see the todo in the function implementation.
   /// \return The fluid density at the givein pose.
-  public: double FluidDensity(const math::Pose3d &_pose) const;
+  public: double UniformFluidDensity(const math::Pose3d &_pose) const;
 
   /// \brief Model interface
   public: Entity world{kNullEntity};
 
   /// \brief The density of the fluid in which the object is submerged in
   /// kg/m^3. Defaults to 1000, the fluid density of water.
   public: double fluidDensity{1000};
+
+  /// \brief When using GradedBuoyancy, we provide a different buoyancy for
+  /// each layer. The key on this map is height in meters and the value is fluid
+  /// density. I.E all the fluid between $key$m and $next_key$m has the density
+  /// $value$kg/m^3. Everything below the first key is considered as having
+  /// fluidDensity.
+  public: std::map<double, double> layers;
 };
 
+
+math::Vector3d getPointOnPlane(
+  math::Plane<double>& plane,
```

Comment:
Style nit, and `const`ness:

```suggestion
  const math::Plane<double> &_plane,
```

The same applies for all functions